lib/node/http.js

```diff
@@ -432,6 +434,133 @@ class HTTP extends Server {
 
       res.json(200, { success: true });
     });
+
+    this.get('/name/:name', async (req, res) => {
+      const valid = Validator.fromRequest(req);
+      const name = valid.str('name');
+
+      if (!name || !rules.verifyName(name))
+        throw new Error('Invalid parameter.');
+
+      const network = this.network;
+      const height = this.chain.height;
+      const nameHash = rules.hashName(name);
+      const reserved = rules.isReserved(nameHash, height + 1, network);
+      const [start, week] = rules.getRollout(nameHash, network);
+      const ns = await this.chain.db.getNameState(nameHash);
+
+      let info = null;
+
+      if (ns) {
+        if (!ns.isExpired(height, network))
+          info = ns.getJSON(height, network);
+      }
+
+      return res.json(200, {
+        start: {
+          reserved: reserved,
+          week: week,
+          start: start
+        },
+        info
+      });
+    });
+
+    this.get('/resource/hash/:hash', async (req, res) => {
+      const valid = Validator.fromRequest(req);
+      const hash = valid.bhash('hash');
+
+      if (!hash)
+        throw new Error('Invalid hash.');
+
+      const ns = await this.chain.db.getNameState(hash);
+      const height = this.chain.tip.height;
+      const network = this.network;
+
+      if (!ns)
+        return res.json(404);
+
+      return res.json(200, ns.getJSON(height,network));
+    });
+
+    this.get('/resource/name/:name', async (req, res) => {
+      const valid = Validator.fromRequest(req);
+      const name = valid.str('name');
+
+      if (!name || !rules.verifyName(name))
+        throw new Error('Invalid name.');
+
+      const nameHash = rules.hashName(name);
+      const ns = await this.chain.db.getNameState(nameHash);
+
+      if (!ns || ns.data.length === 0)
+        return null;
+
+      const resource = Resource.decode(ns.data);
+
+      return res.json(200, resource.getJSON(name));
+    });
+
+    this.get('/proof/name/:name', async (req, res) => {
+      const valid = Validator.fromRequest(req);
+      const name = valid.str('name');
+
+      if (!name || !rules.verifyName(name))
+        throw new Error('Invalid name.');
+
+      const tip = this.chain.tip.hash;
+      const height = this.chain.tip.height;
+      const root = this.chain.tip.treeRoot;
+      const nameHash = rules.hashName(name);
+      const proof = await this.chain.db.prove(root, nameHash);
+
+      return res.json(200, {
+        hash: tip.toString('hex'),
+        height: height,
+        root: root.toString('hex'),
+        name: name,
+        key: nameHash.toString('hex'),
+        proof: proof.toJSON()
```

Comment:
The shape of `proof.toJSON` looks like:

```
  toJSON() {
    return {
      type: typesByVal[this.type],
      depth: this.depth,
      nodes: this.nodes.map(node => node.toJSON()),
      prefix: this.prefix ? this.prefix.toString() : undefined,
      left: this.left ? this.left.toString('hex') : undefined,
      right: this.right ? this.right.toString('hex') : undefined,
      key: this.key ? this.key.toString('hex') : undefined,
      hash: this.hash ? this.hash.toString('hex') : undefined,
      value: this.value ? this.value.toString('hex') : undefined
    };
  }
```

We should try to deduplicate any fields that correspond between the two. It feels a little bit cleaner to me to follow the style in other endpoints where the only object returned is a `primitive.toJSON`.

We want to be 100% sure that we can easily verify the proof client side.

Reply:
:thinking: I dont entirely follow.

This shape mirrors the response of the RPC method:
https://github.com/handshake-org/hsd/blob/master/lib/node/rpc.js#L2375

Would you prefer something like:
```js
    return {
      hash: hash.toString('hex'),
      height: height,
      root: root.toString('hex'),
      name: name,
      key: key.toString('hex'),
      ...proof.toJSON() // <- This would achieve your de-duping.
    };
```

It is definitely cleaner to return _something_.toJSON() - but to confirm, we have other handlers in this file that construct their response "in the handler" like: https://github.com/handshake-org/hsd/blob/master/lib/node/http.js#L354

Whichever we decide - Illl do the same for: https://github.com/handshake-org/hsd/pull/168/files/231e30eef2c442d4914c57cddd840a03624abac9#diff-bbe506ab5582252a30d6872d35fd8ce6R540

Reply:
Current question(s) I _believe_ Im trying to answer:
a. Do we want to return the shape of `proof.toJSON()` with `height` & `root` & `name` included?
or
b. _Something_ else that results in something akin to:
```js
      return res.json(200, primitive.toJSON()); // Assuming just proof.toJSON?
```

Reply:
> but to confirm, we have other handlers in this file that construct their response "in the handler" 

This one was done that way because there was no `filter.JSON` and the filter object lives in a different repo and its hard to coordinate PRs in multiple repos

Reply:
Between a and b, I think something in between is good, it would be nice to have the height and the root along with the rest of `proof.toJSON`. What we really need to be 100% sure of is that we have everything that we need client side to verify the proof, which I believe `proof.toJSON` gives.

Reply:
I added the:
```js
    return {
      hash: hash.toString('hex'),
      height: height,
      root: root.toString('hex'),
      name: name,
      key: key.toString('hex'),
      ...proof.toJSON()
    };
```
:point_up: This feels ok - let me know if you feel otherwise!